### DISCUSSION TITLE
Local Build

### DISCUSSION THREAD

**Discussion by wekempf** [psfinaki/Finelines#7] [Ideas]
One place where you could really provide value over YAML is providing an engine to run the pipeline, verifying it's correctness before committing to version control. I honestly don't know how feasible that is, or how easy to implement, but in my view this is the biggest failing of Azure's pipeline files, above and beyond being YAML.

**Comment by psfinaki** [comment]
Hi @wekempf,

Thanks for your ideas! I've been thinking how to do something like that. 

I mean, committing invalid YAML is a bummer but well at least this things can be validated by running the pipeline from the branch. However it would be cool to not have to commit YAML at all. I have even deliberately asked about it at [StackOverflow](https://stackoverflow.com/questions/68695935/is-there-any-way-to-simulate-dynamic-yaml-pipelines-in-azure-devops/68696201#68696201) but the response is not too optimistic so far.

So yeah I need some ideas on how this can be integrated into CI at least to some extent. If you have any insights, please share!

**Reply by wekempf** [reply]
You'd have to create a build engine, likely working off the source, not the YAML. The hard part is running the "step" when it's using an ADO "plugin". I believe those are mostly just PowerShell scripts, and I *think* they are available for download, but I'm making huge assumptions based on what little I've read about creating these.

**Reply by psfinaki** [reply]
Well if you have some particular links, that will be very helpful. Because writing own build engine doesn't sound like an easy thing to do...
